python/mxnet/gluon/model_zoo/text/lm.py

```diff
@@ -15,110 +15,270 @@
+            begin_state = self.begin_state(batch_size=inputs.shape[1])
+        encoded, state = self.encoder(embedded_inputs, begin_state)
+        out = self.decoder(encoded)
+        return out, state
+
+
+def _get_simple_rnn(embed_dim, hidden_dim, vocab=None,
+                    pretrained=None, ctx=cpu(), root=os.path.join('~', '.mxnet', 'models'),
+                    model_name=None, **kwargs):
+
+    if pretrained:
+        model_name = model_name if not pretrained else '{}_{}'.format(model_name, pretrained)
+        vocab = _load_pretrained_vocab(model_name, root)
+    else:
+        assert vocab is not None, "Must specify vocab if not loading from pretrained models."
+
+    net = SimpleRNN(vocab_size=len(vocab), embed_dim=embed_dim, hidden_dim=hidden_dim,
+                    **kwargs)
+    if pretrained:
+        from ..model_store import get_model_file
+        model_file = get_model_file(model_name, root=root)
+        net.load_params(model_file, ctx=ctx)
+    return net, vocab
+
+
+def _get_awd_rnn(embed_dim, hidden_dim, vocab=None,
+                 pretrained=None, ctx=cpu(), root=os.path.join('~', '.mxnet', 'models'),
+                 model_name=None, **kwargs):
+    if pretrained:
+        model_name = model_name if not pretrained else '{}_{}'.format(model_name, pretrained)
+        vocab = _load_pretrained_vocab(model_name, root)
+    else:
+        assert vocab is not None, "Must specify vocab if not loading from pretrained models."
+
+    net = AWDRNN(vocab_size=len(vocab), embed_dim=embed_dim, hidden_dim=hidden_dim, **kwargs)
+    if pretrained:
+        from ..model_store import get_model_file
+        model_file = get_model_file(model_name, root=root)
+        net.load_params(model_file, ctx=ctx)
+    return net, vocab
+
+
+def awd_lstm_lm_1150(**kwargs):
+    r"""3-layer LSTM language model with weight-drop, variational dropout, and tied weights.
+
+    Embedding size is 400, and hidden layer size is 1150.
+
+    Parameters
+    ----------
+    vocab : gluon.text.Vocabulary, default None
+        Vocabulary object to be used with the language model.
+        Required when not loading from pretrained models.
+    pretrained : str or None, default None
+        The dataset name on which the pretrained model is trained. Options are 'wikitext2'.
+        If None, then no pretrained weights are loaded.
+    ctx : Context, default CPU
+        The context in which to load the pretrained weights.
+    root : str, default '~/.mxnet/models'
+        Location for keeping the model parameters.
+
+    Returns
+    -------
+    Block, gluon.text.Vocabulary
+    """
+    predefined_args = {'model_name': 'awd_lstm_lm_1150',
+                       'mode': 'lstm',
+                       'num_layers': 3,
+                       'tie_weights': True,
+                       'dropout': 0.4,
+                       'weight_drop': 0.5,
+                       'drop_h': 0.3,
+                       'drop_i': 0.65}
```

Comment:
I feel that it's not very clear to store the predefined_args in a separate dictionary. We can directly use
```python
def awd_lstm_lm_1150(model_name=..., mode=..., ...)
```

Reply:
This provides the opportunity for user to override some settings, which complicates how we manage pretrained models. For such need, they can instantiate from the class directly without involving the functions. These functions are only intended as serving model with a fixed set of hyperparameters, and pretrained weights.

Reply:
We still need a function to define the preset hyperparameters. I can put that in a dictionary.
For creating AWD, I think the constructor of the classes already serve that purpose.